ratatool-shapeless/src/main/scala/com/spotify/ratatool/shapeless/CaseClassDiffy.scala

```diff
@@ -0,0 +1,178 @@
+  MapEncoder[FieldType[K, Double]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.self)
+    }
+  }
+
+  implicit val hnilEncoder: MapEncoder[HNil] =
+    createEncoder[HNil](n => Map(HNil.toString -> HNil))
+
+  implicit def seqEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Seq[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.toIndexedSeq)
+    }
+  }
+
+  implicit def listEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, List[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.toIndexedSeq)
+    }
+  }
+
+  /*
+  * null.asInstanceOf[V] derives the default value of primitive types
+  * null.asInstanceOf[Int] = 0
+  * null.asInstanceOf[Boolean] = false
+  * null.asInstanceOf[String] = null
+  * */
+  implicit def optionEncoder[K <: Symbol, V](implicit witness: Witness.Aux[K]):
+  MapEncoder[FieldType[K, Option[V]]] = {
+    val name = witness.value.name
+    createEncoder { v =>
+      Map(name -> v.getOrElse(null.asInstanceOf[V]))
+    }
+  }
+
+  //scalastyle:off
+  implicit def hlistEncoder0[K <: Symbol, H, T <: HList](implicit hEncoder: Lazy[MapEncoder[FieldType[K, H]]],
+                                                          tEncoder: Lazy[MapEncoder[T]]): MapEncoder[FieldType[K, H] :: T] = {
+    createEncoder[FieldType[K, H] :: T](in => hEncoder.value.toMap(in.head) ++ tEncoder.value.toMap(in.tail))
+  }
+
+  implicit def hListEncoder1[K <: Symbol, H, T <: HList, R <: HList](implicit
+                                                                      wit: Witness.Aux[K],
+                                                                      gen: LabelledGeneric.Aux[H, R],
+                                                                      encoderH: Lazy[MapEncoder[R]],
+                                                                      encoderT: Lazy[MapEncoder[T]]
+                                                                     ): MapEncoder[FieldType[K, H] :: T] =
+    createEncoder(in =>
+      encoderT.value.toMap(in.tail) ++ Map(wit.value.name -> encoderH.value.toMap(gen.to(in.head))))
+
+  implicit def genericEncoder[A, R](implicit gen: LabelledGeneric.Aux[A, R],
+                                    enc: MapEncoder[R]): MapEncoder[A] = {
+    createEncoder(a => enc.toMap(gen.to(a)))
+  }
+}
+
+class CaseClassDiffy[T](ignore: Set[String] = Set.empty,
+                        unordered: Set[String] = Set.empty)(implicit diff: MapEncoder[T])
+  extends Diffy[T](ignore, unordered) {
+  import CaseClassDiffy._
+
+  override def apply(x: T, y: T): Seq[Delta] = {
+    diff(Some(asMap(x)), Some(asMap(y)))
+      .filter(f => !hnilIgnore.exists(f.field.contains(_)))
+  }
+
+  private def hnilIgnore = ignore ++ Set(HNil.toString)
+
+  private def diff(left: Any, right: Any, pref: String = "")
+  : Seq[Delta] = {
```

Comment:
Nit, can this be moved onto the same line as function def?

Reply:
👍